### ISSUE TITLE
Mapping to an existing FHIR profiles [High Priority]

### ISSUE THREAD

**Issue by markkramerus** [standardhealth/shr-fhir-export#53]
To assure compatibility of SHR with existing sets of profiles, I would like to build off of those profiles, rather than base FHIR resources. For example,  PersonOfRecord could be mapped to US-Core Patient, rather than the base FHIR Patient. Since they are both StructureDefinitions, this should be possible, but I don't know if it is currently supported.  How would I say that in the mapping file?

PersonOfRecord maps to http://hl7.org/fhir/us/core/StructureDefinition/us-core-patient:

**Comment by cmoesel** [comment]
In theory, I don't *think* this would be too difficult -- as long as we have the JSON definitions of the profiles you wish to map to.  I'll take a look at it today to see what might be involved.

**Comment by cmoesel** [comment]
Question: If an SHR profile is mapped to a US Core profile, we have two possible ways to implement it.  For example, if PersonOfRecord maps to us-core-patient, we could:
1. Create a new profile with `baseDefinition: "http://hl7.org/fhir/us/core/StructureDefinition/us-core-patient"`, and the differential only contains differences from the US Core profile.
2. Create a new profile with `baseDefinition: "http://hl7.org/fhir/StructureDefinition/Patient"` and the differential contains all differences from FHIR Patient resource (US Core and SHR differences).

The first case would be _formally_ extending the US Core profile, whereas the second case would not formally extend the US Core profile, but would be a US Core profile clone plus SHR modifications.  There is some risk in the first case, in that I'm not sure that there are many profiles on profiles out there yet -- we may be pioneering new territory.

**Comment by cmoesel** [comment]
I've implemented most of this according to approach 2 above.  I've still not yet fully determined the best way to deal with merging mappings into existing slices in the target profile... but I have a few ideas.

One thing that just struck me, however, is this situation:
```
Observation maps to Observation:
// bunch of field mappings

BodyWeight maps to http://hl7.org/fhir/StructureDefinition/bodyweight
// few specific mappings (maybe)
```

Currently, you only inherit mappings from a parent if the child doesn't declare a new target, or if the child's target is the same as the parent's (e.g., parent maps to Observation and child maps to Observation).

In the example above, the child's target is technically different than the parent's, so the current implementation would _not_ inherit the parent mappings.  _But..._ the child's target is a profile on the parent's target, so perhaps the parent's specific mappings _should_ still be inherited.  I *think* they probably should, but would appreciate the opinions of others (particularly @markkramerus).

**Comment by markkramerus** [comment]
I was really thinking/hoping for approach #1. I could have sworn I gave you that guidance, but maybe I replied to the email notification accidentally.  Both approaches should give the same snapshot, but in the first approach the consistency with the base profile is more "provable".

**Comment by markkramerus** [comment]
On the field inheritance with a profiled "based on"... The problem is that we are conflating mapping with constraint inheritance. Mapping is how fields are translated to fields. Mapping to a profile means the resulting snapshot ALSO inherits the constraints from the target. So in answer to your question, the field mappings SHOULD be inherited from Observation, AND the additional extensions and constraints of the base profile SHOULD be inherited from the profiled base class.

**Comment by markkramerus** [comment]
How's that blog post coming?

**Comment by cmoesel** [comment]
I hadn't received any indication of preference for # 1 vs. # 2 -- so perhaps you (@markkramerus)  did respond to the automated Github message and it went into the ether.  I went toward # 2, because I hadn't ever seen _anyone_ do # 1.  That said, I'll modify to take approach # 1 then.

As for inheriting field mappings from the parent mappings to Observation, I'll work on that as well.  It's a bit tricky because that inheritance gets materialized in a component of the tool chain that doesn't have access to the profile definition (so it doesn't know what it should inherit from).  But I'll figure something out.

As for the blog post, that's currently prioritized below getting the mapping to profiles working.  Let me know if that's the wrong priority.

**Comment by markkramerus** [comment]
If I'm off base on #1 versus #2 then tell me!  It makes me nervous when you say no one does it that way. Plus, I don't want to waste time zigging and zagging. I think we need to discuss it in more detail.

**Comment by cmoesel** [comment]
You're not off-base, @markkramerus !  Since then I've done some more looking around and found that the approach we call # 1 (basing a profile on a profile) is actually employed in FHIR's own vital signs profiles.  There is a parent "VitalsSigns" profile, and then all specific vital signs profiles (e.g., bp, body weight, etc) are based on it.  Prior to this I hadn't seen it anywhere, but if it's good enough for the FHIR core team, it's good enough for me!

**Comment by cmoesel** [comment]
This was released in `shr-fhir-export` v5.2.0.
